<a href="https://colab.research.google.com/github/divik544/FashionMNIST/blob/master/fashion_mnist_CNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
!pip install -U tensorflow_datasets

Requirement already up-to-date: tensorflow_datasets in /usr/local/lib/python3.6/dist-packages (1.0.2)


In [0]:
import tensorflow as tf
import tensorflow_datasets as tfds
import math
import numpy as np
import matplotlib.pyplot as plt
tf.enable_eager_execution()
tf.logging.set_verbosity(tf.logging.ERROR)

### Load Dataset

In [0]:
dataset,metadata = tfds.load('fashion_mnist',as_supervised=True,with_info=True)
train_ds,test_ds = dataset['train'],dataset['test']

In [10]:
num_train = metadata.splits['train'].num_examples
num_test = metadata.splits['test'].num_examples
print(num_train)

60000


### Preprocess the data

Normalize the value of pixels.


In [0]:
def normalize(image,label):
  image = tf.cast(image,tf.float32)
  image /= 255
  return image,label


train_ds = train_ds.map(normalize)
test_ds = test_ds.map(normalize)

### Build the Model


In [0]:
model = tf.keras.Sequential([
    tf.keras.layers.Conv2D(32,(3,3),padding='same',activation=tf.nn.relu,input_shape=(28,28,1)),
    tf.keras.layers.MaxPooling2D((2,2),strides=2),
    tf.keras.layers.Conv2D(64,(3,3),padding='same',activation=tf.nn.relu),
    tf.keras.layers.MaxPooling2D((2,2),strides=2),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128,activation=tf.nn.relu),
    tf.keras.layers.Dense(10,activation=tf.nn.softmax)
])

In [0]:
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])

In [0]:
# specify iteration parameters
BATCH_SIZE = 32
train_ds = train_ds.repeat().shuffle(num_train).batch(BATCH_SIZE)
test_ds = test_ds.batch(BATCH_SIZE)

In [13]:
model.fit(train_ds,epochs=10,steps_per_epoch=math.ceil(num_train/BATCH_SIZE))

Epoch 1/10
1875/1875 [==============================] - 26s 14ms/step - loss: 0.1311 - acc: 0.9508
Epoch 2/10
1875/1875 [==============================] - 17s 9ms/step - loss: 0.1155 - acc: 0.9573
Epoch 3/10
1875/1875 [==============================] - 16s 9ms/step - loss: 0.0973 - acc: 0.9642
Epoch 4/10
1875/1875 [==============================] - 17s 9ms/step - loss: 0.0800 - acc: 0.9704
Epoch 5/10
1875/1875 [==============================] - 17s 9ms/step - loss: 0.0685 - acc: 0.9752
Epoch 6/10
1875/1875 [==============================] - 18s 9ms/step - loss: 0.0591 - acc: 0.9789
Epoch 7/10
1875/1875 [==============================] - 17s 9ms/step - loss: 0.0512 - acc: 0.9811
Epoch 8/10
1875/1875 [==============================] - 16s 9ms/step - loss: 0.0434 - acc: 0.9836
Epoch 9/10
1875/1875 [==============================] - 17s 9ms/step - loss: 0.0387 - acc: 0.9861
Epoch 10/10
1875/1875 [==============================] - 18s 10ms/step - loss: 0.0342 - acc: 0.9870


In [15]:
test_loss,test_accuracy = model.evaluate(test_ds,steps=math.ceil(num_test/BATCH_SIZE))
print(test_accuracy)

313/313 [==============================] - 2s 7ms/step - loss: 0.4052 - acc: 0.9178
0.9178


###Results

Accuracy on training data : 98.7%

Accuracy on test data: 91.8%